<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111: Introduction to Databases
</span>
</center></b>
</span><br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Lecture 2: Begin Working with Data Modeling, the Relational Model and SQL<br>(V 2.1)
</span>
</center></i>

# Introduction

## Questions?

## Today's Topics

1. Environment: Spend some time configuring environments and discussing what is happening:
    - Jupyter (iPython) Notebook
    - URLs, database connections, database management system servers
    - PyCharm environments
<br><br>
2. Core database concepts and examples.
<br><br>
3. The relational model and SQL:
    1. Overview and concepts.
    2. SQL operations: ```SELECT, INSERT, UPDATE, DELETE.```
    3. Introduction to keys.
    4. Discussion of HW1.
    

# Set Up Environment

## Setup

In [26]:
%load_ext sql
%sql mysql+pymysql://dbuser:dbuserdbuser@localhost/lahman2019raw
%sql select * from people where playerid='willite01'

import pymysql
default_cnx = pymysql.connect(host='localhost',
                             user='dbuser',
                             password='dbuserdbuser',
                             db='lahman2019raw',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


- To run the examples that use %sql, you will have to install an extension to your iPython/Jupyter notebooks.


- The extension and installation instruction are at: https://pypi.org/project/ipython-sql/


- You can come to office hours (CAs, mine) if you need help.

| <img src="../../images/install_wizard.jpg"> |
| :---: |
| __Installation Wizard__ |

## Explanation

- Three interesting and illuminating program statements.

```
%load_ext sql
%sql mysql+pymysql://dbuser:dbuser@localhost/lahman2017
%sql select * from people where playerid='willite01'
```
<br><br>

| <img src="../../images/what-is-this-sorcery.jpg" > |
| :---: |
| __Some Sorcery__ |

### iPython/Jupyter Notebook Architecture

| <img src="../../images/notebook_components.png" > |
| :---: |
| [iPython Notebook Architecture](https://jupyter.readthedocs.io/en/latest/architecture/how_jupyter_ipython_work.html) |


- There are several good explanations of interactive Python, iPython notebooks, etc:
    - https://jupyter.readthedocs.io/en/latest/architecture/how_jupyter_ipython_work.html
    - https://ipython-books.github.io/chapter-3-mastering-the-jupyter-notebook/
    
    
- You do not need to understand this material for the course.


- In lectures, I am using Jupyter notebooks.
    - There is an ecosystem of plug-ins and extensions that add features to the base notebook capabilities.
    - The ```%sql magic function``` is one of the extensions. 
    
    
- "IPython will treat any line whose first character is a % as a special call to a ‘magic’ function. These allow you to control the behavior of IPython itself, plus a lot of system-type features. They are all prefixed with a % character, but parameters are given without parentheses or quote." (https://ipython.readthedocs.io/en/stable/interactive/reference.html#magic)

In [3]:
%magic

- ```%load_ext sql``` is calling the magic function ```load_extension.``` This
    - Loads the extension module
    - The ```%sql``` magic function.
  
  
- This is all FYI. The [iPython](https://ipython.readthedocs.io/en/stable/config/custommagics.html?highlight=magic) documentation explains how to implement Magics.     

### Databases and Connections

- What about ```%sql mysql+pymysql://dbuser:dbuser@localhost/lahman2017```?

| <img src="../../images/db-connection.jpg"> |
| :---: |
| __Database Server and Connection__ |


- There are typically two processes (running programs) in a database application.
    - The database server, which processes database commands.
    - The client application that sends commands to and receives responses from the database server.
    
    
- The magic statement defines a database connection.


"A Database connection is facility in computer science that allows client software to talk to database server software, whether on the same machine or not. A connection is required to send commands and receive answers, usually in the form of a result set.

Connections are a key concept in data-centric programming. Since some DBMS engines require considerable time to connect connection pooling was invented to improve performance. No command can be performed against a database without an "open and available" connection to it.

Connections are built by supplying an underlying driver or provider with a connection string, which is a way of addressing a specific database or server and instance as well as user authentication credentials (for example, Server=sql_box;Database=Common;User ID=uid;Pwd=password;). Once a connection has been built it can be opened and closed at will, and properties (such as the command time-out length, or transaction, if one exists) can be set. The Connection String is composed of a set of key/value pairs as dictated by the data access interface and data provider being used." (https://en.wikipedia.org/wiki/Database_connection)

- Modifying the diagram above to add the DB server

| <img src="../../images/notebook_and_db.jpeg"> |
| :---: |
| __Notebook and DB__

- Again, this is just FYI.
    - Understanding that there are connections, how to create them, ... is important and you will write code that uses connections.
    - We will not cover the details of connections.

### URLs

- A little more about ```mysql+pymysql://dbuser:dbuser@localhost/lahman2017```


- The connection specification above is a URL.


- "A Uniform Resource Locator (URL), colloquially termed a web address, is a reference to a web resource that specifies its location on a computer network and a mechanism for retrieving it. A URL is a specific type of Uniform Resource Identifier (URI), although many people use the two terms interchangeably. URLs occur most commonly to reference web pages (http), but are also used for file transfer (ftp), email (mailto), database access (JDBC), and many other applications." (https://en.wikipedia.org/wiki/URL)


- A URL has the format


```URI = scheme:[//authority]path[?query][#fragment]```


- The components are:
    1. Scheme: Information about the protocol, connector library, ...
    2. Authority: Usually ```userid:password```.
    3. Path: File system like folder path to the resource.
    4. We will cover query string later.
    5. Fragment: A location or subset of the resource, e.g. a section with heading.
    
    
- We have seen how we connect to MySQL from Python using ```PyMySQL``` library.

```default_cnx = pymysql.connect(host='localhost',
                             user='dbuser',
                             password='dbuser',
                             db='lahman2017',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)```



- Some connector libraries support a single connection string of the form:


```jdbc:mysql://someuserid:somepassword@www.myurl.com:3306```


- Other databases have similar concepts, for example CouchDB.


### The Command

- And, ```%sql select * from people where playerid='willite01'``` simply runs a [Structured Query Language (SQL)](https://en.wikipedia.org/wiki/SQL) statement.


- The optional (recommended) text book _Ramakrishnan and Gehrke,_ chapter 5 discusses SQL.


- SQL is the main topic in this and the next few lectures.

### URL, Connection Summary

- A database management system:
    - Is a software program executing on a "server" in the network.The data in a database.
    - Stores the data in the database on the file system, usually in special files on a "disk."
    
    
- Applications, including tools like MySQL Workbench, connect to the database management system:
    - Over a network.
    - Specifying some form of URL that defines:
        - The network address of the server (host, port).
        - The protocol to use.
        - The specific database among the many that the DBMS controls.
        - Authentication and authorization information.
    - Send commands in a DBMS specific language to the server.

| <img src="../../images/two-three-tier.jpeg"> |
| :---: |
| __Database Architectures__ |

- We have seen an example of (a) using Jupyter Notebooks.

In [10]:
import json
cool_cnx = pymysql.connect(host='localhost',
                             user='dbuser',
                             password='dbuserdbuser',
                             db='lahman2019raw',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

sql = "select playerID, nameLast, nameFirst, birthCity, birthState, birthYear from People where birthCity='San Diego' " + \
    "and nameLast='Williams'"
cur = cool_cnx.cursor()
res = cur.execute(sql)
people_info = cur.fetchall()
print("The relevant people are: \n", json.dumps(people_info, indent=2))

The relevant people are: 
 [
  {
    "playerID": "willite01",
    "nameLast": "Williams",
    "nameFirst": "Ted",
    "birthCity": "San Diego",
    "birthState": "CA",
    "birthYear": "1918"
  },
  {
    "playerID": "willitr01",
    "nameLast": "Williams",
    "nameFirst": "Trevor",
    "birthCity": "San Diego",
    "birthState": "CA",
    "birthYear": "1992"
  }
]


- Other database have similar concepts, for example [CouchDB](http://couchdb.apache.org/)

| <img src="../../images/couchdb_arch.png" width="800px"> |
| :---: |
| [CouchDB Architecture](https://www.researchgate.net/figure/Simple-Architecture-of-Apache-CouchDB-database-Storage-Engine-It-is-B-tree-based-and-the_fig1_265062016) |

In [23]:
import requests
body = {
    "selector": {
        "last_name": {"$eq": "Ferguson"}
    },
    "fields": ["_id", "_rev", "last_name", "email"]
}
url = "http://127.0.0.1:5984/w4111f19/_find"
headers = {"content-type": "application/json"}
result = requests.post(url, json=body, headers=headers)
print("Result = ", json.dumps(result.json(), indent=2))


Result =  {
  "docs": [
    {
      "_id": "2570e0b648805195a380643653006e66",
      "_rev": "1-b2d6f5b544ea86d95522f3af40d0e2c5",
      "last_name": "Ferguson",
      "email": "dff9@columbia.edu"
    }
  ],
  "bookmark": "g1AAAABweJzLYWBgYMpgSmHgKy5JLCrJTq2MT8lPzkzJBYorGJmaG6QaJJmZWFgYmBpamiYaWxiYmRibmRobGJilmpmB9HHA9BGlIwsAJ70bdw",
  "warning": "no matching index found, create an index to optimize query time"
}


- We will do a Web simple application, specifically a REST API, the implements the three-tier model. This will be homework 2.


- There is a very basic HTML UI to give a feel for how the application works end-to-end. The primary focus is an application that accesses data.

In [28]:
from IPython.display import IFrame  
IFrame('http://127.0.0.1:5000/static/W4111/index.html', width=900, height=650)


In [30]:
%rest http://127.0.0.1:5000/api/lahman2019raw/people/willite01

{
  "data": {
    "playerID": {
      "value": "aardsda01",
      "link": {
        "rel": "people",
        "href": "/api/people/aardsda01"
      }
    },
    "birthYear": "1981",
    "birthMonth": "12",
    "birthDay": "27",
    "birthCountry": "USA",
    "birthState": "CO",
    "birthCity": "Denver",
    "deathYear": "",
    "deathMonth": "",
    "deathDay": "",
    "deathCountry": "",
    "deathState": "",
    "deathCity": "",
    "nameFirst": "David",
    "nameLast": "Aardsma",
    "nameGiven": "David Allan",
    "weight": "215",
    "height": "75",
    "bats": "R",
    "throws": "R",
    "debut": "2004-04-06",
    "finalGame": "2015-08-23",
    "retroID": "aardd001",
    "bbrefID": "aardsda01"
  },
  "links": [
    {
      "rel": "current",
      "href": "http://127.0.0.1:5000/api/lahman2019raw/people/willite01"
    }
  ]
}

<Response [200]>

## PyCharm (Python) Environment

### Concept

- A core concept in Python, and by extension PyCharm and any Python tool, is a virtual environment.


"'What Is a Virtual Environment?'"<br><br>
At its core, the main purpose of Python virtual environments is to create an isolated environment for Python projects. This means that each project can have its own dependencies, regardless of what dependencies every other project has." (https://realpython.com/python-virtual-environments-a-primer/)

- [From the Python documentation:](https://docs.python.org/3/tutorial/venv.html)

"Python applications will often use packages and modules that don’t come as part of the standard library. Applications will sometimes need a specific version of a library, because the application may require that a particular bug has been fixed or the application may be written using an obsolete version of the library’s interface.

This means it may not be possible for one Python installation to meet the requirements of every application. If application A needs version 1.0 of a particular module but application B needs version 2.0, then the requirements are in conflict and installing either version 1.0 or 2.0 will leave one application unable to run.

The solution for this problem is to create a virtual environment, a self-contained directory tree that contains a Python installation for a particular version of Python, plus a number of additional packages.

Different applications can then use different virtual environments. To resolve the earlier example of conflicting requirements, application A can have its own virtual environment with version 1.0 installed while application B has another virtual environment with version 2.0. If application B requires a library be upgraded to version 3.0, this will not affect application A’s environment."


- Other programming languages have similar concepts:
    - Java Classpath (https://en.wikipedia.org/wiki/Classpath_(Java))
    - JavaScript/NodeJS has the [Node Module Path](https://nodejs.org/api/modules.html)
    
    
- There are two applications with environments that I will use in this course.
    - Ananconda
    - PyCharm
    
### PyCharm Demo

- Demo environments, creating environments, installing packages.


### Anaconda

- Demonstrate terminal environment in Jupyter.


- Demonstrate Conda command prompt.

# Common Database Concepts

## The Concepts

- Almost all database engines and models have the concepts of
    - Objects that are some form of array of (name, value) pairs.
    - Sets of similar or related objects.
    - Four basic (CRUD) operations on a set
        - CREATE a new object and add to a set.
        - RETRIEVE an object in a set based on a criteria.
        - UPDATE an object in a set, e.g. change the data in the object.
        - DELETE an object from a set, specifying the object(s) by some criteria.
        
        
- In the file systems/CSV model:
    - A set is a file, e.g. People.csv.
    - Each object is a row in the file.
    - The header row gives the names of each column.
    - The CRUD processing involves writing a program that reads the file, changes the two-dimensional array and writing the file.
        - CREATE: Append a row and save the file.
        - RETRIEVE: Scan the table and apply some kind of IF statement.
        - UPDATE: Change a row in the two dimensional array.
        - DELETE: Remove a row from the array.
        

- In the "pure" relational model
    - A set is a _relation_.
    - An object is a _row_ or _tuple_.
    - The operation for _retrieve_ is $\sigma$.
    - There is no support for CREATE, UPDATE or DELETE.
    - There is an _algebra_ and language for producing a new relation from existing relations that you can use to "sort of" implement INSERT, DELETE, UPDATE.
    
    
- In SQL,
    - A set is a _table_.
    - An object is a _row_ or _tuple_.
    - INSERT is the create operation.
    - UPDATE is the delete operation.
    - DELETE is the delete operation.
    - SELECT is the statement that realizes the relational _algebra_.
    
    
- In [Representational state transfer](https://en.wikipedia.org/wiki/Representational_state_transfer) REST.
    - A set is a _resource_ that is a collection of _resources._
    - An object is a resource.
    - CREATE is HTTP POST
    - RETRIEVE is HTTP GET
    - UPDATE is HTTP PUT (or PATCH)
    - DELETE is HTTP DELETE.
    
    
- [Redis](https://redis.io/): I am tempted to say, "Don't ask!"
    - The basic concept is that Redis stores pairs of (key, value).
    - Key is just a string
    - Value can be:
        - String
        - List
        - Hash
        - Set
        - Sorted Set
    - The CRUD commands depend on the data type for a value.
    
    
- In [CouchDB](http://couchdb.apache.org/): I am tempted to say, "No. Seriously, do not ask! Please!"
    - CouchDB is basically a REST/HTTP accessible document database.
    - The CRUD operations are the same as for basic REST, but ...
    - There is much more powerful support for running commands, e.g. ```_find.```
    

## Examples

### SQL

- ```RETRIEVE``` is ```SELECT```


- The basic syntax is:
```
SELECT column1, column2, ...
FROM table_name
WHERE condition;
```

- ```table_name``` is the table/relation representing a set of entities.


- ```column1, column2, ...``` is the list of fields/properties you want from each entity.


- ```WHERE condition``` is like an ```IF``` statement in a programming language. The operation only returns the entities for which the condition is true.

In [11]:
%%sql
    select playerID, nameLast, nameFirst, birthYear, birthState 
    from lahman2019raw.people where nameLast='Williams' and birthState='CA'
    

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
11 rows affected.


[('willibe01', 'Williams', 'Bernie', '1948', 'CA'),
 ('willido02', 'Williams', 'Don', '1935', 'CA'),
 ('williji03', 'Williams', 'Jimy', '1943', 'CA'),
 ('willike02', 'Williams', 'Ken', '1964', 'CA'),
 ('willima04', 'Williams', 'Matt', '1965', 'CA'),
 ('willimi02', 'Williams', 'Mitch', '1964', 'CA'),
 ('williri02', 'Williams', 'Rinaldo', '1893', 'CA'),
 ('williri03', 'Williams', 'Rick', '1952', 'CA'),
 ('willish01', 'Williams', 'Shad', '1971', 'CA'),
 ('willite01', 'Williams', 'Ted', '1918', 'CA'),
 ('willitr01', 'Williams', 'Trevor', '1992', 'CA')]

- Remember, we set up a connection. So, the "location" in the "web" of the entity is ```localhost:3038``` which resolve to my laptop. 


### REST

- SQL is very powerful and there is a widely implement [standard.](https://en.wikipedia.org/wiki/ISO/IEC_9075)


- REST is less a standard and more a set of conventions or informal agreements.


- Different endpoints interpret the conventions differently. You typically have to read documentation for each endpoint to figure it all out.


- For HW2, we will implement a REST endpoint on top of the Lahman 2019 database.


- A sample query would be:
```
http://127.0.0.1:5000/api/lahman2019raw/people?nameLast=Williams&birthState=CA&fields=playerID,nameLast,nameFirst,birthYear,birthState
```

- A sample of using my solution to HW2 is:

In [1]:
 %load_ext restmagic

In [7]:
%rest GET http://127.0.0.1:5000/api/lahman2019raw/people?nameLast=Williams&birthState=CA&fields=playerID,nameLast,nameFirst,birthYear,birthState

{
  "data": [
    {
      "playerID": {
        "value": "willibe01",
        "link": {
          "rel": "people",
          "href": "/api/people/willibe01"
        }
      },
      "nameLast": "Williams",
      "nameFirst": "Bernie",
      "birthYear": "1948",
      "birthState": "CA"
    },
    {
      "playerID": {
        "value": "willido02",
        "link": {
          "rel": "people",
          "href": "/api/people/willido02"
        }
      },
      "nameLast": "Williams",
      "nameFirst": "Don",
      "birthYear": "1935",
      "birthState": "CA"
    },
    {
      "playerID": {
        "value": "williji03",
        "link": {
          "rel": "people",
          "href": "/api/people/williji03"
        }
      },
      "nameLast": "Williams",
      "nameFirst": "Jimy",
      "birthYear": "1943",
      "birthState": "CA"
    },
    {
      "playerID": {
        "value": "willike02",
        "link": {
          "rel": "people",
          "href": "/api/people/willike02"
        

<Response [200]>

- Comments:
    - 127.0.0.1 is just another way to say ```localhost.```
    - ```/api``` is a convention or pattern that means "I am calling an API, not asking for a web page.
    - ```/lahman2019/people``` is a convention for identifying the database and table.
    - ```fields=...``` is a convention for identifying the properties that you want.
    - ```?x=y&z=m``` is a simplified query expressions similar to ```WHERE ...```
    
    
- We will understand more when we move onto HW2. The important point is thinking about the web as a kind of database.


- __Note:__ Demo the UI


<hr style="height:1px;">

| <img src="../../images/hw2_intro.jpeg"> |
| :---: |
| __HW2 Concept__ |

<hr style="height:1px;">

### Files/CSVs

- There is not standard. HW1 defines a convention, and then maps to RDB or CSV files.


- To learn about database concepts, we are implementing a DB like think on top of CSV files.


- We implement the same concept on top of RDB to compare implementations.


- We could do the same pattern on top of all of our various databases, but there is not a lot of learning value.

In [8]:
# Load my solution implementation into the Notebook's path.
import sys
sys.path.append("C:\\Users\\dferguso\\ansys_one_drive\\Columbia\\Projects\\W4111\\HW1Impl")

In [12]:
from src.CSVDataTable import CSVDataTable

In [13]:
# Where is the "DB" holding my "tables"?
_default_directory = \
        "C:\\Users\\dferguso\\ansys_one_drive\\Columbia\Projects\\W4111\\HW1Impl\\Data" + \
            "\\baseballdatabank-2019.2\\core\\" 

In [14]:
# I want to connect to a "table"
connect_info = {
    "directory": _default_directory,
    "file_name": 'People.csv'
}

In [15]:
# Create a data table to wrap the file.
t = CSVDataTable(table_name="People", key_columns=['playerID'],connect_info=connect_info)                     

In [16]:
# Load the file into memory.
t.load()

In [18]:
template = {"birthState": "CA", "nameLast": "Williams"}
field_list = ['playerID', 'nameLast', 'nameFirst', 'birthYear', 'birthState']

In [19]:
result = t.find_by_template(template, field_list, None, None, None)

In [20]:
print(result)

<class 'src.DerivedDataTable.DerivedDataTable'>: name = FBT:People
connect_info = None
Key columns = None
No. of rows = 11
{'playerID': 'willibe01', 'nameLast': 'Williams', 'nameFirst': 'Bernie', 'birthYear': '1948', 'birthState': 'CA'}
{'playerID': 'willido02', 'nameLast': 'Williams', 'nameFirst': 'Don', 'birthYear': '1935', 'birthState': 'CA'}
{'playerID': 'williji03', 'nameLast': 'Williams', 'nameFirst': 'Jimy', 'birthYear': '1943', 'birthState': 'CA'}
{'playerID': 'willike02', 'nameLast': 'Williams', 'nameFirst': 'Ken', 'birthYear': '1964', 'birthState': 'CA'}
{'playerID': 'willima04', 'nameLast': 'Williams', 'nameFirst': 'Matt', 'birthYear': '1965', 'birthState': 'CA'}


### Redis, CouchDB

__I said, 'Don't ask."__


- These are complex discussions and I want to move onto the relational model.

# The Relational Model $-$ I

## Overview

References: Ramakrishnan and Gehrke, section 3.1

There are two perspectives on the relational model:
1. Formal language and algebra.
2. A standard implementation language, Structure Query Language [(SQL)](https://en.wikipedia.org/wiki/SQL)


The relational model and SQL language (and most DB engines) have two sub-languages/models:
1. Data definition
2. Data manipulation
    
    
MySQL and other database engines also have an API/language for administration and security.

<hr style="height:1px;">

| <img src="../../images/db_engine_apis.jpeg"> |
| :---: |
| __DB Engine Interfaces__ |

<hr style="height:1px;">

<hr>

| <img src="../../images/Slide03.jpg"> |
| :---: |
| __Relational Model__ |

<hr>

| <img src="../../images/Slide04.jpg"> |
| :---: |
| __Relational Model__ |

## Defining Data

### Relational Model

- _Relational Schema_ defines and provides metadata describing what is/can be in "the file" (called a _relation_), e.g.<br>


\begin{equation}
Employees(\overline{employeeNumber}:integer, \ lastName:string, \ firstName:string, \ extension:string, \ email:string, \ officeCode:integer, \ reportsTo:integer, \ jobTitle: string)
\end{equation}


- The formal relational model is imprecise about allowed types for fields/columns. The types must be elementary (atomic) types, e.g. integer, string, char, ...


- The $\overline{overline}$ indicates the _key columns,_ which we will cover later.


- A _Relational Instance_ is a "table" of data that conforms to a relation schema, e.g.<br><br>


| <img src="../../images/employee_relation.jpeg" width="900px%"> |
| :---: |
| __Employee Relational Instance__ |

### SQL

- Structured Query Language (SQL) is a well-defined programming language that realizes a superset of the relation model.


- There are two sub-languages
    1. Data Definition Language (DDL) corresponds to relational schema.
    2. Data Manipulation Language corresponds to relational algebra (covered later).
    

- Data Definition Statements (DDL) statements are how you create/modify tables in SQL.


- Defining a table in SQL uses the create statement. The statement for creating the ```People``` table to hold data from ```Lahman2019's``` CSV file ```people.csv``` is: 

```

CREATE TABLE `people` (
  `playerID` text,
  `birthYear` text,
  `birthMonth` text,
  `birthDay` text,
  `birthCountry` text,
  `birthState` text,
  `birthCity` text,
  `deathYear` text,
  `deathMonth` text,
  `deathDay` text,
  `deathCountry` text,
  `deathState` text,
  `deathCity` text,
  `nameFirst` text,
  `nameLast` text,
  `nameGiven` text,
  `weight` text,
  `height` text,
  `bats` text,
  `throws` text,
  `debut` text,
  `finalGame` text,
  `retroID` text,
  `bbrefID` text
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;



```

- The back, single quotes are a sometimes useful/sometimes annoying aspect of MySQL's implementation of the SQL standard.
    - Anything in between single back quotes is a _schema element._
    - Allows attributes and table names of the form \`My Table\` \`My-Table\` and makes clear that \`My-Table\` is a schema element and not an attempt to subtract the variable `Table` from the variable `My` 


- SQL has a superset of the relational model's schema definition capability. An improved table definition could be
        
        
- There are some design or best practice issues with this table, which we will discuss later.
        
        
- The SQL DML
    - Has many other operations/statements.
        - DROP
        - ALTER
    - Operates on more than just a table's attributes, e.g.
        - Constraints
        - Triggers
        - Indexes


- We will go into DML and defining data in much more detail.

## Manipulating and Processing Information

### Computer Languages

- A [declarative language's](https://en.wikipedia.org/wiki/Declarative_programming) programs describe their desired results without explicitly listing commands or steps that must be performed


- An [imperative language](https://en.wikipedia.org/wiki/Imperative_programming) program consists of commands for the computer to perform. Imperative programming focuses on describing how a program operates.


- The most distinguishing feature is that declarative languages do not have control-flow constructs, e.g.
    - ```if {...} else {...}```
    - ```for (...) {...}```


- The relation model, SQL and many DB languages are (primarily) [declarative languages.](https://en.wikipedia.org/wiki/Declarative_programming)

### Algebraic Query Language

Reference: Ramkrishan and Gherke, section 2.4.

Both the abstract relational model/theory and SQL are __algebras__.

#### Algebra

"... abstract algebra (occasionally called modern algebra) is the study of algebraic structures." (https://en.wikipedia.org/wiki/Abstract_algebra) 

"In mathematics, and more specifically in abstract algebra, an algebraic structure is a set (called carrier set or underlying set) with one or more operations defined on it that satisfies a list of axioms." (https://en.wikipedia.org/wiki/Algebraic_structure)

__Group is an Example__ (http://mathworld.wolfram.com/Group.html)

"A group G is a finite or infinite set of elements together with a binary operation (called the group operation) that together satisfy the four fundamental properties of closure, associativity, the identity property, and the inverse property. The operation with respect to which a group is defined is often called the "group operation," and a set is said to be a group "under" this operation. Elements A, B, C, ... with binary operation between A and B denoted AB form a group if

1. Closure: If A and B are two elements in G, then the product AB is also in G.
<br><br>
2. Associativity: The defined multiplication is associative, i.e., for all A,B,C in G, (AB)C=A(BC).
<br><br>
3. Identity: There is an identity element I (a.k.a. 1, E, or e) such that IA=AI=A for every element A in G.
<br><br>
4. Inverse: There must be an inverse (a.k.a. reciprocal) of each element. Therefore, for each element A of G, the set contains an element B=A<sup>-1</sup> such that <br>AA<sup>-1</sup>=A<sup>-1</sup>A=I.

#### Why a Special Query Language? Why a Formal Language? Why an Algebra

- Computing has a formal mathemetical model.


- Programming languages derive from the model and have their own formal definition.


- Almost every time someone publishes a new language, my reaction is, "What? Why another language? Can't we just pick one, use it and get it right?"


- Relational algebra is less powerful and expressive than Java, C, ... and other programming languages.


- __The simplicity and constrained capabilities is actually one of the core of the value,__ enabling
    - Vastly simplified programming and supporting tools that yields increased productivity.
    - Development of algorithms that process the data definitions and query statements to automatically produce optimal execution plans, which are better than what a programmer can directly code.
    
    
- We will see these benefits in coming lectures.


_Simple Tool Example_ that enables "citizen programmers" to maniuplate data.

| <img src="../../images/ss_1.jpeg"> |
| :---: |
| __Data Processing and Cleansing Tools__ |

<br><br>
_Simple Tool Example_ for enabling business professionals to analyze and report on data.

| <img src="../../images/quicksight.jpeg" width="85%"> |
| :---: |
| [QuickSight](https://aws.amazon.com/quicksight/) Analysis Tool |

### Relational Algebra

- There are two notations or representations of the algebra:
    - The original, formal theory.
    - SQL
    
#### Original, Formal Relational Notation

Ramakrishnan and Gehrke, section 4.2


- The "set" in the relational algebra is the set of _relations_.


- The operations are:
    - Common set operations:
        - Union: $\cup$
        - Intersection: $\cap$
        - Difference: $-$
    - Projection: $\pi$
    - Selection: $\sigma$
    - Cartesian Product: $\times$
    - Join: $\bowtie$
    - Rename/Alias
    

- The formal notation does not support create, update or delete of tuples/rows in a relation. You could emulate the operations by
    - Defining new relations containing the created, updated or delete tuples.
    - Using $\cup$, $\cap$, $-$ on the original relation and created/deleted/updated tuple relation.
    
    
- This is an algebra $\Rightarrow$ that $R_1\ OP\ R_2 \rightarrow R_3 ,$ where $R_1 , R_2 , $ and $R_3$ are relations ( aka tables). Relationa algebra is closed. Algebra statements operate on relations and produce relations.
    
    
#### SQL Notation

Ramakrishnan and Gehrke, sections 5.1, 5.2

- The "set" in the relational algebra is the catalog of _tables_.


- The operations are:
    - Common set operations:
        - Union: UNION
        - Intersection: INTERSECT
        - Difference: EXCEPT
    - Selection, Projection, Cartesian Product, Rename/Alias and Join are clauses within a SELECT statement.
    
    ```SELECT <project clause> FROM <table> [JOIN <table> [ON <join condition]] WHERE <select condition>```
    
    
- SQL supports additional operations, e.g
    - GROUP BY
    - ORDER BY
    


### Selection and Projection

#### Relational Model

- <span style="color: red;">Projection</span> produces a new relation that
    - Has the same rows as the original table
    - But only containing the requested columns/fields
    
    
- <span style="color: red;">Selection</span> produces a new relation that
    - Has the same columns as the original relation
    - But only contains rows with column values matching a predicate.
    

##### Selection

- The selection operator is $\sigma$


- The selection predicate operators are: $\lt, \gt, =, \ne, \ge, \le$


- $\sigma(Players)$ selects all of the rows/tuples in the player relation.


- The predicate/condition is a "subscript," e.g.
    - $\sigma$<sub>playerID=napolmi01</sub>$(Players)$ selects all players with _playerID=napolmi01_
    - $\sigma$<sub>((nameLast=Williams)$\land(throws=L))\lor(birthYear\ne1914)$</sub>$(Players)$ selects all players with
        - Last name williams who threw lefthanded
        - Birth year not equal to 1914
        
        
- This notation is a little clunky. You can think of the selection predicate being similar to what goes in an _if()_ statement if you were looping through an array testing for a match.


###### Projection

- The projection operator is $\pi$


- The requested columns are subscripts on the operator, e.g. $\pi$<sub>$nameLast,nameFirst,throws$</sub>$(Players)$ returns a table
    - Containing the $nameLast, nameFirst, throws$ column values, in that order
    - For all tuples in the $Players$ table.

##### An Algebra

- Because the relational model is an algebra, composition is possible.


- Like algebras, there are instances where composition supports _commutativity,_ _associativity,_ _distribution_ and other algebra rules. 


- For example, $\pi$ and $\sigma$ are commutative.
\begin{equation}
\pi_{nameLast,nameFirst,birthYear}(\sigma_{birthYear=1960}(Players)) = \sigma_{birthYear=1960}(\pi_{nameLast,nameFirst,birthYear}(Players))
\end{equation}


- The fact that some algebra rules apply will become important in query optimization later in the course.

#### SQL

- The core syntax of the SQL SELECT statement is
```
SELECT [DISTINCT] column-list
    FROM table-list
    WHERE qualification
```

- All RDB products implement the core syntax but with extensions.


- The MYSQL syntax is

```
SELECT
    [ALL | DISTINCT | DISTINCTROW ]
      [HIGH_PRIORITY]
      [STRAIGHT_JOIN]
      [SQL_SMALL_RESULT] [SQL_BIG_RESULT] [SQL_BUFFER_RESULT]
      SQL_NO_CACHE [SQL_CALC_FOUND_ROWS]
    select_expr [, select_expr ...]
    [FROM table_references
      [PARTITION partition_list]
    [WHERE where_condition]
    [GROUP BY {col_name | expr | position}, ... [WITH ROLLUP]]
    [HAVING where_condition]
    [WINDOW window_name AS (window_spec)
        [, window_name AS (window_spec)] ...]
    [ORDER BY {col_name | expr | position}
      [ASC | DESC], ... [WITH ROLLUP]]
    [LIMIT {[offset,] row_count | row_count OFFSET offset}]
    [INTO OUTFILE 'file_name'
        [CHARACTER SET charset_name]
        export_options
      | INTO DUMPFILE 'file_name'
      | INTO var_name [, var_name]]
    [FOR {UPDATE | SHARE} [OF tbl_name [, tbl_name] ...] [NOWAIT | SKIP LOCKED] 
      | LOCK IN SHARE MODE]]
```


- __Examples are the easiest way to master the syntax and behavior.__


- A reminder of the ```Lahman2017.People``` table.

```
CREATE TABLE `people` (
  `playerID` varchar(12) NOT NULL,
  `birthYear` char(4) DEFAULT NULL,
  `birthMonth` char(2) DEFAULT NULL,
  `birthDay` char(2) DEFAULT NULL,
  `birthCountry` varchar(64) DEFAULT NULL,
  `birthState` varchar(32) DEFAULT NULL,
  `birthCity` varchar(64) DEFAULT NULL,
  `deathYear` char(4) DEFAULT NULL,
  `deathMonth` char(2) DEFAULT NULL,
  `deathDay` char(2) DEFAULT NULL,
  `deathCountry` varchar(64) DEFAULT NULL,
  `deathState` varchar(32) DEFAULT NULL,
  `deathCity` varchar(64) DEFAULT NULL,
  `nameFirst` varchar(64) NOT NULL,
  `nameLast` varchar(64) NOT NULL,
  `nameGiven` varchar(128) DEFAULT NULL,
  `weight` char(3) DEFAULT NULL,
  `height` char(3) DEFAULT NULL,
  `bats` char(1) DEFAULT NULL,
  `throws` char(1) DEFAULT NULL,
  `debut` varchar(16) DEFAULT NULL,
  `finalGame` varchar(16) DEFAULT NULL,
  `retroID` varchar(12) DEFAULT NULL,
  `bbrefID` varchar(12) DEFAULT NULL,
  PRIMARY KEY (`playerID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

- __NOTE:__ All of the single back quotes are:
    - MySQL specific
    - Optional in most cases, and only required if a table or column name has a space or some other icky character, e.g. ``` `First Name` ```, ```First-Name```

<u>Select all People with last name 'Williams'</u>

In [8]:
%sql select * from People where nameLast='Williams';

80 rows affected.


[('williac01', 1917, 3, 18, 'USA', 'NJ', 'Montclair', 1999, 9, 16, 'USA', 'FL', 'Fort Myers', 'Ace', 'Williams', 'Robert Fulton', '174', '74', 'R', 'L', '1940-07-15', '1946-04-22', 'willa103', 'williac01', datetime.date(1999, 9, 16), datetime.date(1917, 3, 18)),
 ('willial02', 1914, 5, 11, 'USA', 'AL', 'Valhermosa Springs', 1969, 7, 19, 'USA', 'TX', 'Groves', 'Al', 'Williams', 'Almon Edward', '200', '75', 'R', 'R', '1937-04-19', '1938-09-04', 'willa101', 'willial02', datetime.date(1969, 7, 19), datetime.date(1914, 5, 11)),
 ('willial03', 1954, 5, 6, 'Nicaragua', 'Atlantico Sur', 'Pearl Lagoon', None, None, None, '', '', '', 'Albert', 'Williams', 'Albert Hamilton', '190', '76', 'R', 'R', '1980-05-07', '1984-09-26', 'willa001', 'willial03', None, datetime.date(1954, 5, 6)),
 ('williar01', 1877, 8, 24, 'USA', 'MA', 'Somerville', 1941, 5, 16, 'USA', 'VA', 'Arlington', 'Art', 'Williams', 'Arthur Frank', '', '', 'L', 'R', '1902-05-07', '1902-09-01', 'willa104', 'williar01', datetime.date(1941, 5, 16), datetime.date(1877, 8, 24)),
 ('willibe01', 1948, 10, 8, 'USA', 'CA', 'Alameda', None, None, None, '', '', '', 'Bernie', 'Williams', 'Bernard', '175', '73', 'R', 'R', '1970-09-07', '1974-05-12', 'willb105', 'willibe01', None, datetime.date(1948, 10, 8)),
 ('willibe02', 1968, 9, 13, 'P.R.', '', 'San Juan', None, None, None, '', '', '', 'Bernie', 'Williams', 'Bernabe', '180', '74', 'B', 'R', '1991-07-07', '2006-10-01', 'willb002', 'willibe02', None, datetime.date(1968, 9, 13)),
 ('willibi01', 1938, 6, 15, 'USA', 'AL', 'Whistler', None, None, None, '', '', '', 'Billy', 'Williams', 'Billy Leo', '175', '73', 'L', 'R', '1959-08-06', '1976-10-02', 'willb104', 'willibi01', None, datetime.date(1938, 6, 15)),
 ('willibi02', 1932, 6, 13, 'USA', 'SC', 'Newberry', 2013, 6, 11, 'USA', 'CA', 'Berkeley', 'Billy', 'Williams', 'William', '195', '75', 'L', 'R', '1969-08-15', '1969-08-20', 'willb102', 'willibi02', datetime.date(2013, 6, 11), datetime.date(1932, 6, 13)),
 ('willibo01', 1884, 4, 27, 'USA', 'OH', 'Monday', 1962, 8, 6, 'USA', 'OH', 'Nelsonville', 'Bob', 'Williams', 'Robert Elias', '190', '72', 'R', 'R', '1911-07-03', '1913-09-30', 'willb103', 'willibo01', datetime.date(1962, 8, 6), datetime.date(1884, 4, 27)),
 ('willibr01', 1969, 2, 15, 'USA', 'SC', 'Lancaster', None, None, None, '', '', '', 'Brian', 'Williams', "Brian O'Neal", '205', '75', 'R', 'R', '1991-09-16', '2000-07-05', 'willb001', 'willibr01', None, datetime.date(1969, 2, 15)),
 ('willich01', 1947, 10, 11, 'USA', 'NY', 'Flushing', 2015, 1, 27, 'USA', 'FL', 'Port Orange', 'Charlie', 'Williams', 'Charles Prosek', '200', '74', 'R', 'R', '1971-04-23', '1978-09-11', 'willc101', 'willich01', datetime.date(2015, 1, 27), datetime.date(1947, 10, 11)),
 ('willicy01', 1887, 12, 21, 'USA', 'IN', 'Wadena', 1974, 4, 23, 'USA', 'WI', 'Eagle River', 'Cy', 'Williams', 'Fred', '180', '74', 'L', 'L', '1912-07-18', '1930-09-22', 'willc103', 'willicy01', datetime.date(1974, 4, 23), datetime.date(1887, 12, 21)),
 ('willida01', 1855, 10, 6, 'USA', 'KY', 'Ludlow', 1939, 10, 22, 'USA', 'KY', 'Covington', 'Dale', 'Williams', 'Elisha Alphonso', '175', '69', 'R', 'R', '1876-08-12', '1876-09-09', 'willd106', 'willida01', datetime.date(1939, 10, 22), datetime.date(1855, 10, 6)),
 ('willida02', 1881, 2, 7, 'USA', 'PA', 'Scranton', 1918, 4, 25, 'USA', 'AR', 'Hot Springs', 'Dave', 'Williams', 'David Owen', '167', '71', 'R', 'L', '1902-07-02', '1902-08-03', 'willd103', 'willida02', datetime.date(1918, 4, 25), datetime.date(1881, 2, 7)),
 ('willida03', 1927, 11, 2, 'USA', 'TX', 'Dallas', 2009, 8, 17, 'USA', 'TX', 'Farmers Branch', 'Davey', 'Williams', 'David Carlous', '160', '70', 'R', 'R', '1949-09-16', '1955-07-31', 'willd102', 'willida03', datetime.date(2009, 8, 17), datetime.date(1927, 11, 2)),
 ('willida05', 1958, 2, 28, 'USA', 'NY', 'Brooklyn', None, None, None, '', '', '', 'Dallas', 'Williams', 'Dallas McKinley', '165', '71', 'L', 'L', '1981-09-19', '1983-10-02', 'willd101', 'willida05', None, datetime.date(19

- I am using a special plug-in [ipython-sql](https://pypi.org/project/ipython-sql/) for running SQL directly from a code cell in the Jupyter notebook.


- I can use the SQL command from many database clients.

| <img src="../../images/sql_query_1.jpeg"> |
| :---: |
| __Query MySQL Workbench__ |

| <img src="../../images/sql_select_terminal_1.jpeg"> |
| :---: |
| __Query MySQL Terminal Client__ |

- From Python code


In [31]:
import pymysql.cursors
import pandas as pd
import json

# The database server is running somewhere in the network.
# I must specify the IP address (HW server) and port number
# (connection that SW server is listening on)
# Also, I do not want to allow anyone to access the database
# and different people have different permissions. So, the
# client must log on.


# Connect to the database over the network. Use the connection
# to send commands to the DB.
default_cnx = pymysql.connect(host='localhost',
                             user='dbuser',
                             password='dbuser',
                             db='lahman2017',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)


def run_q(q, args=None, fields= None, fetch=True, cnx=None):
    """

    :param q: An SQL query string that may have %s slots for argument insertion.
    :param args: A tuple of values to insert in the %s slots.
    :param fetch: If true, return the result.
    :param cnx: A database connection. May be None
    :return: A result set or None.
    """

    try:
        if cnx is None:
            cnx = default_cnx

        if fields:
            q = q.format(",".join(fields))

        cursor=cnx.cursor()             # Just ignore this for now.

        print("Query = ", cursor.mogrify(q, args))
        cursor.execute(q, args)         # Execute the query.

        if fetch:
            r = cursor.fetchall()       # Return all elements of the result.
        else:
            r = None
    except Exception as e:
        print("Exception e = ", e)
        
    cnx.commit()

    return r

In [32]:
result = run_q("select * from people where nameLast='Williams';")
print("Result = ")
print(result)

Query =  select * from people where nameLast='Williams';
Result = 
[{'playerID': 'williac01', 'birthYear': 1917, 'birthMonth': 3, 'birthDay': 18, 'birthCountry': 'USA', 'birthState': 'NJ', 'birthCity': 'Montclair', 'deathYear': 1999, 'deathMonth': 9, 'deathDay': 16, 'deathCountry': 'USA', 'deathState': 'FL', 'deathCity': 'Fort Myers', 'nameFirst': 'Ace', 'nameLast': 'Williams', 'nameGiven': 'Robert Fulton', 'weight': '174', 'height': '74', 'bats': 'R', 'throws': 'L', 'debut': '1940-07-15', 'finalGame': '1946-04-22', 'retroID': 'willa103', 'bbrefID': 'williac01', 'DOD': datetime.date(1999, 9, 16), 'DOB': datetime.date(1917, 3, 18)}, {'playerID': 'willial02', 'birthYear': 1914, 'birthMonth': 5, 'birthDay': 11, 'birthCountry': 'USA', 'birthState': 'AL', 'birthCity': 'Valhermosa Springs', 'deathYear': 1969, 'deathMonth': 7, 'deathDay': 19, 'deathCountry': 'USA', 'deathState': 'TX', 'deathCity': 'Groves', 'nameFirst': 'Al', 'nameLast': 'Williams', 'nameGiven': 'Almon Edward', 'weight': '200

- OK. That output is kind of icky.


- The Python function json.dumps() makes the output a little cleaner.

<u>A more complex example that includes PROJECT</u>

In [33]:
q ='''SELECT nameLast as last_name, nameFirst as firstName, birthYear as birth_year  
    from people where 
    nameLast='Williams' and birthYear > '1920';'''
result = run_q(q)
print("Result = ")
print(json.dumps(result, indent=2))

Query =  SELECT nameLast as last_name, nameFirst as firstName, birthYear as birth_year  
    from people where 
    nameLast='Williams' and birthYear > '1920';
Result = 
[
  {
    "last_name": "Williams",
    "firstName": "Albert",
    "birth_year": 1954
  },
  {
    "last_name": "Williams",
    "firstName": "Bernie",
    "birth_year": 1948
  },
  {
    "last_name": "Williams",
    "firstName": "Bernie",
    "birth_year": 1968
  },
  {
    "last_name": "Williams",
    "firstName": "Billy",
    "birth_year": 1938
  },
  {
    "last_name": "Williams",
    "firstName": "Billy",
    "birth_year": 1932
  },
  {
    "last_name": "Williams",
    "firstName": "Brian",
    "birth_year": 1969
  },
  {
    "last_name": "Williams",
    "firstName": "Charlie",
    "birth_year": 1947
  },
  {
    "last_name": "Williams",
    "firstName": "Davey",
    "birth_year": 1927
  },
  {
    "last_name": "Williams",
    "firstName": "Dallas",
    "birth_year": 1958
  },
  {
    "last_name": "Williams",
    "f

<u>Using Arguments</u>

- If you want to pass parameters into a query string, there are two options:
    1. String concatentation functions.
    2. %s templates and arguments


- You cannot insert schema terms, like table name or columns using %s substitution. So, you using string formatting or concatenation.


In [34]:
q = "select {} from {} where nameLast=%s and throws=%s"
q2 = q.format(",".join(['playerId', 'nameLast', 'nameFirst', 'throws']), "People")
print("q2 = ", q2)

q2 =  select playerId,nameLast,nameFirst,throws from People where nameLast=%s and throws=%s


In [35]:
cur = default_cnx.cursor()
cur.execute(q2, ('Williams', 'R'))
result = cur.fetchall()
print("Result = ", json.dumps(result, indent=2))

Result =  [
  {
    "playerId": "willial02",
    "nameLast": "Williams",
    "nameFirst": "Al",
    "throws": "R"
  },
  {
    "playerId": "willial03",
    "nameLast": "Williams",
    "nameFirst": "Albert",
    "throws": "R"
  },
  {
    "playerId": "williar01",
    "nameLast": "Williams",
    "nameFirst": "Art",
    "throws": "R"
  },
  {
    "playerId": "willibe01",
    "nameLast": "Williams",
    "nameFirst": "Bernie",
    "throws": "R"
  },
  {
    "playerId": "willibe02",
    "nameLast": "Williams",
    "nameFirst": "Bernie",
    "throws": "R"
  },
  {
    "playerId": "willibi01",
    "nameLast": "Williams",
    "nameFirst": "Billy",
    "throws": "R"
  },
  {
    "playerId": "willibi02",
    "nameLast": "Williams",
    "nameFirst": "Billy",
    "throws": "R"
  },
  {
    "playerId": "willibo01",
    "nameLast": "Williams",
    "nameFirst": "Bob",
    "throws": "R"
  },
  {
    "playerId": "willibr01",
    "nameLast": "Williams",
    "nameFirst": "Brian",
    "throws": "R"
  },
  

## A Digression on the Power of SQL

- CourseWorks (Canvass) has a basic REST API for querying and updating information about courses, students, ...


- More complex understanding of the data requires:
    - Making GET calls to retrieve the information using query parameters.
    - Writing a program to process the data to answer complex questions.
    
    
- SQL provides simpler (once you understand it), more compact ways to analyze data than writing Python, Java, etc. programs.


- To show with an example,
    - I wrote a Python program to query and download information from CourseWorks/Canvass.
    - Create and load the data into MySQL tables.
    
    
- What does the data look like?

<hr style="height:1px;">

| <img src="../../images/cw_er_model.jpg"> |
| :---: |
| __My CourseWorks Data Model__ |

<hr style="height:1px;">

- The notation in the diagram is an example of [Crow's Foot Notation.](https://www.vertabelo.com/blog/technical-articles/crow-s-foot-notation).


- We will go through the meanings, patterns, etc. later.


- The CourseWorks REST API allows me to query what classes I have taken/taught, which students have been in which classes, etc.


- I cannot ask interesting questions about complex combinations of the data.


- <u>Question 1: "Who is the bravest person in the world?"</u> The answer is the person who has taken the most courses with me.

In [25]:
%%sql
select last_name, first_name, count(*) as number_of_courses from w4111f19.student_course_view
    group by student_id
    having number_of_courses > 2
    order by count(*) desc
    limit 5;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


[('Gupta', ' Abhi', 4), ('Liu', ' Sen', 3), ('Ontiveros', ' Bryan', 3)]

- I will not bore you with more Python code, but doing this on the REST APIs alone would be tedious.


- The RDB query engine executes the diagram below to produce the result. You can think of each rectangle as being a block of code that you would have to write.


<hr style="height:1px;">

| <img src="../../images/query_execution.jpg" width="700px;"> |
| :---: |
| __Query Execution Plan__ |

<hr style="height:1px;">

- <u>Question 1: "Who is in this class and has previously taken a class with me? Maybe they can help you."</u>

In [27]:
%%sql
use w4111f19;

SELECT
	first_name, count(*) as count, group_concat(course_code_full separator ", ") as course_codes_full 
FROM
	student_course_view
where
	student_id in ( select student_id from student_course_view where course_id='85777' or course_id='91884')
    and
    student_id in (select student_id from student_course_view where year < 2019 or (semester < 3 and year = 2019))
group by
	user_id, last_name, first_name
having count > 1
order by count desc;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
0 rows affected.
9 rows affected.


[(' Samantha', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Sharon', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Ahmed', 2, 'ENGIE1102_001_2018_3 - THE ART OF ENGINEERING, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Joshua', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Bradley', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Derek', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Jetasri', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Jillian', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES'),
 (' Lara', 2, 'ENGIE1006_001_2018_1 - INTRO TO COMP FOR ENG/APP SCI, COMSW4111_002_2019_3 - INTRODUCTION TO DATABASES')]

- That SQL statement may seem complex until you realize the equivalent code algorithm is something like ...

<hr style="height:1px;">

| <img src="../../images/explain_2.png" width="900px;"> |
| :---: |
| __Query Execution Plan__ |

<hr style="height:1px;">

## Connections, Cursors and Fetch...

- We previously, briefly discussed connections.


- Section 6.1.2 in Ramakrishnan and Gherke discuss cursors (p. 189, 191)


"SQL is a set-based language, meaning operations are completed on all or rows of the result.  However, there are times, when you want to do operation on a row by row basis.  This is where cursors come in to play.

... ...

A database cursor can be thought of as a pointer to a specific row within a query result.  The pointer can be moved from one row to the next.  Depending on the type of cursor, you may be even able to move it to the previous row." (https://www.essentialsql.com/database-cursor/)


- Cursors are fundamental to relational database APIs, and many other types of databases.


- The details are often hidden and not explicit, and programmers use them in advanced scenarios. But, they are always (logically) created. 


| <img src="../../images/cursor.png"> |
| :---: |
| [Cursor Concept](https://codeandwork.github.io/courses/java/sp_cursors_triggers_indexes.html) |


- Concept is similar to an iterator in programming languages/libraries.


- ```cursor.execute(...)``` sends the command to the database and produces the result set. The common operations on the result set are:
    - ```fetchall()```
    - ```fetchone()```
    - ```fetchmany(size=...)```
    
    
- ```cursor.mogrify(q, args)``` returns the actual query string that will be sent to the database.


- The details are PyMySQL library specific but the concepts apply in general to all connection libraries.


- A simple example:

In [30]:
import pymysql
default_cnx = pymysql.connect(host='localhost',
                             user='dbuser',
                             password='dbuserdbuser',
                             db='lahman2019raw',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

In [43]:
cur = default_cnx.cursor()
how_many = cur.execute("select * from people where nameLast='Smith'")
print("The query produced", how_many, "rows.")

The query produced 162 rows.


In [44]:
import json
all_result = cur.fetchall()
print("Fetch all returns a list of size: ", len(all_result))
print("The result set is = \n", json.dumps(all_result, indent=2))

Fetch all returns a list of size:  162
The result set is = 
 [
  {
    "playerID": "smith01",
    "birthYear": "",
    "birthMonth": "",
    "birthDay": "",
    "birthCountry": "",
    "birthState": "",
    "birthCity": "",
    "deathYear": "",
    "deathMonth": "",
    "deathDay": "",
    "deathCountry": "",
    "deathState": "",
    "deathCity": "",
    "nameFirst": "",
    "nameLast": "Smith",
    "nameGiven": "",
    "weight": "",
    "height": "",
    "bats": "",
    "throws": "",
    "debut": "1884-06-05",
    "finalGame": "1884-06-05",
    "retroID": "smitu101",
    "bbrefID": "smith01"
  },
  {
    "playerID": "smithal01",
    "birthYear": "1871",
    "birthMonth": "",
    "birthDay": "",
    "birthCountry": "USA",
    "birthState": "NY",
    "birthCity": "New York",
    "deathYear": "1919",
    "deathMonth": "7",
    "deathDay": "9",
    "deathCountry": "USA",
    "deathState": "NY",
    "deathCity": "New York",
    "nameFirst": "Aleck",
    "nameLast": "Smith",
    "nameGiven

In [49]:
cur = default_cnx.cursor()
how_many = cur.execute("select * from people where nameLast='Smith'")
print("The query produced", how_many, "rows.")

The query produced 162 rows.


In [50]:
result = cur.fetchmany(size=3)
while result is not None and len(result) > 0:
    print("The next 3 rows are:\n", json.dumps(result, indent=2))
    cont = input("Do you want to see more (Y/N)?")
    if cont == 'Y' or cont=='y':
        result = cur.fetchmany(size=3)
    else:
        result = None

The next 3 rows are:
 [
  {
    "playerID": "smith01",
    "birthYear": "",
    "birthMonth": "",
    "birthDay": "",
    "birthCountry": "",
    "birthState": "",
    "birthCity": "",
    "deathYear": "",
    "deathMonth": "",
    "deathDay": "",
    "deathCountry": "",
    "deathState": "",
    "deathCity": "",
    "nameFirst": "",
    "nameLast": "Smith",
    "nameGiven": "",
    "weight": "",
    "height": "",
    "bats": "",
    "throws": "",
    "debut": "1884-06-05",
    "finalGame": "1884-06-05",
    "retroID": "smitu101",
    "bbrefID": "smith01"
  },
  {
    "playerID": "smithal01",
    "birthYear": "1871",
    "birthMonth": "",
    "birthDay": "",
    "birthCountry": "USA",
    "birthState": "NY",
    "birthCity": "New York",
    "deathYear": "1919",
    "deathMonth": "7",
    "deathDay": "9",
    "deathCountry": "USA",
    "deathState": "NY",
    "deathCity": "New York",
    "nameFirst": "Aleck",
    "nameLast": "Smith",
    "nameGiven": "Alexander Benjamin",
    "weight": 

- There are many implications for cursors. Two of the most important are:
    1. Remember that the database application is a network connected client of the database. Returning all rows in a large result may swamp the client program or network.
    2. DBMS use locking to implement transactions, which we will cover later. Cursors are fundamental to locking.

## INSERT, UPDATE, DELETE

### Note on Examples

I do not want to "mess up" the Lahman2017 stats with insert, update, delete. So, I am going to use the the [MySQL Tutorial Classic Cars](http://www.mysqltutorial.org/mysql-sample-database.aspx) database/schema.

| <img src="http://www.mysqltutorial.org/wp-content/uploads/2009/12/MySQL-Sample-Database-Schema.png"> |
| :---: |
| __Classic Cars Schema__ |

### Overview

- All databases support some form of Create-Retrieve-Update-Delete (CRUD)


- ```SELECT``` is the SQL retrieve operation.


- The verbs for the other operations are:
    - ```INSERT``` for Create
    - ```UPDATE```
    - ```DELETE```
    
### INSERT

#### Overview

- The basic form is

```
INSERT INTO table_name (column1, column2, column3, ...)
VALUES (value1, value2, value3, ...);
```

- If you are passing a value for every column, in the order in which the columns are defined in the table, you can skip the columns list because it is implied.

```
INSERT INTO table_name
VALUES (value1, value2, value3, ...);
```

- For MySQL and most RDB engines, the ```INSERT``` statement has many more options. The MySQL version is

```
INSERT [LOW_PRIORITY | DELAYED | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [(col_name [, col_name] ...)]
    {VALUES | VALUE} (value_list) [, (value_list)] ...
    [ON DUPLICATE KEY UPDATE assignment_list]

INSERT [LOW_PRIORITY | DELAYED | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    SET assignment_list
    [ON DUPLICATE KEY UPDATE assignment_list]

INSERT [LOW_PRIORITY | HIGH_PRIORITY] [IGNORE]
    [INTO] tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [(col_name [, col_name] ...)]
    SELECT ...
    [ON DUPLICATE KEY UPDATE assignment_list]

value:
    {expr | DEFAULT}

value_list:
    value [, value] ...

assignment:
    col_name = value

assignment_list:
    assignment [, assignment] ...
```

- We will not cover most of the options in the course.

#### Examples

__Note:__ We will switch to a simpler table to demonstrate ```INSERT, UPDATE``` and ```DELETE.```

In [54]:
%sql select * from classicmodels.offices

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
7 rows affected.


[('1', 'San Francisco', '+1 650 219 4782', '100 Market Street', 'Suite 300', 'CA', 'USA', '94080', 'NA'),
 ('2', 'Boston', '+1 215 837 0825', '1550 Court Place', 'Suite 102', 'MA', 'USA', '02107', 'NA'),
 ('3', 'NYC', '+1 212 555 3000', '523 East 53rd Street', 'apt. 5A', 'NY', 'USA', '10022', 'NA'),
 ('4', 'Paris', '+33 14 723 4404', "43 Rue Jouffroy D'abbans", None, None, 'France', '75017', 'EMEA'),
 ('5', 'Tokyo', '+81 33 224 5000', '4-1 Kioicho', None, 'Chiyoda-Ku', 'Japan', '102-8578', 'Japan'),
 ('6', 'Sydney', '+61 2 9264 2451', '5-11 Wentworth Avenue', 'Floor #2', None, 'Australia', 'NSW 2010', 'APAC'),
 ('7', 'London', '+44 20 7877 2041', '25 Old Broad Street', 'Level 7', None, 'UK', 'EC2N 1HN', 'EMEA')]

In [12]:
%%sql
    INSERT INTO classicmodels.offices
    values(8, 'New York', '+1 212 555-1212', 'W 116th Street', 'and Broadway', 'NY', 'USA', '10027', 'NA')

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[]

In [57]:
%sql select * from classicmodels.offices

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
8 rows affected.


[('1', 'San Francisco', '+1 650 219 4782', '100 Market Street', 'Suite 300', 'CA', 'USA', '94080', 'NA'),
 ('2', 'Boston', '+1 215 837 0825', '1550 Court Place', 'Suite 102', 'MA', 'USA', '02107', 'NA'),
 ('3', 'NYC', '+1 212 555 3000', '523 East 53rd Street', 'apt. 5A', 'NY', 'USA', '10022', 'NA'),
 ('4', 'Paris', '+33 14 723 4404', "43 Rue Jouffroy D'abbans", None, None, 'France', '75017', 'EMEA'),
 ('5', 'Tokyo', '+81 33 224 5000', '4-1 Kioicho', None, 'Chiyoda-Ku', 'Japan', '102-8578', 'Japan'),
 ('6', 'Sydney', '+61 2 9264 2451', '5-11 Wentworth Avenue', 'Floor #2', None, 'Australia', 'NSW 2010', 'APAC'),
 ('7', 'London', '+44 20 7877 2041', '25 Old Broad Street', 'Level 7', None, 'UK', 'EC2N 1HN', 'EMEA'),
 ('8', 'New York', '+1 212 555-1212', 'W 116th Street', 'and Broadway', 'NY', 'USA', '10027', 'NA')]

- Example insert with explicitly enumerated columns:

In [13]:
%%sql
    INSERT INTO classicmodels.offices
    (city, officeCode, country, phone, territory, addressLine1, postalCode)
    values('Boston', 9, 'USA', '+1 617 555 1212', 'NA', '1 Government Plaza', '02101')

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[]

In [14]:
%sql select * from classicmodels.offices

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
9 rows affected.


[('1', 'San Francisco', '+1 650 219 4782', '100 Market Street', 'Suite 300', 'CA', 'USA', '94080', 'NA'),
 ('2', 'Boston', '+1 215 837 0825', '1550 Court Place', 'Suite 102', 'MA', 'USA', '02107', 'NA'),
 ('3', 'NYC', '+1 212 555 3000', '523 East 53rd Street', 'apt. 5A', 'NY', 'USA', '10022', 'NA'),
 ('4', 'Paris', '+33 14 723 4404', "43 Rue Jouffroy D'abbans", None, None, 'France', '75017', 'EMEA'),
 ('5', 'Tokyo', '+81 33 224 5000', '4-1 Kioicho', None, 'Chiyoda-Ku', 'Japan', '102-8578', 'Japan'),
 ('6', 'Sydney', '+61 2 9264 2451', '5-11 Wentworth Avenue', 'Floor #2', None, 'Australia', 'NSW 2010', 'APAC'),
 ('7', 'London', '+44 20 7877 2041', '25 Old Broad Street', 'Level 7', None, 'UK', 'EC2N 1HN', 'EMEA'),
 ('8', 'New York', '+1 212 555-1212', 'W 116th Street', 'and Broadway', 'NY', 'USA', '10027', 'NA'),
 ('9', 'Boston', '+1 617 555 1212', '1 Government Plaza', None, None, 'USA', '02101', 'NA')]

### UPDATE

#### Overview

- The basic syntax is:

```
UPDATE table_name
    SET column1 = value1, column2 = value2, ...
    WHERE condition;
```

- The ```WHERE``` condition syntax and behavior is exactly the same as the ```WHERE``` clause in a ```SELECT.```


- There ```WHERE``` clause determines the matching rows. The ```SET``` changes the values of the specified columns.


- The full MySQL syntax is a little more complicated.

```
UPDATE [LOW_PRIORITY] [IGNORE] table_reference
    SET assignment_list
    [WHERE where_condition]
    [ORDER BY ...]
    [LIMIT row_count]

value:
    {expr | DEFAULT}

assignment:
    col_name = value

assignment_list:
    assignment [, assignment] ...
```

#### Examples

##### Example 1: Simple

- Change of the zip code for the office with officeCode=11.

In [77]:
%sql select * from classicmodels.offices where officeCode=9

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[('9', 'Boston', '+1 617 555 1212', '1 Government Plaza', None, None, 'USA', '02101', 'NA')]

In [78]:
%sql update classicmodels.offices set postalCode="02343",city="Pembroke" where officeCode=9

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[]

In [79]:
%sql select * from classicmodels.offices where officeCode=9

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[('9', 'Pembroke', '+1 617 555 1212', '1 Government Plaza', None, None, 'USA', '02343', 'NA')]

##### Example 2: More Complex

- Increase the MSRP for products from ```productLine``` Motorcycles and ```productScale``` 1:10 by 5%.


- First, let's get a feel for the ```products``` table.

In [2]:
%sql select count(*) as productCount from classicmodels.products;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[(110,)]

- There are 110 products. Let's just look at the first 10 to understand structure.

In [3]:
%sql select * from classicmodels.products limit 10;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
10 rows affected.


[('S10_1678', '1969 Harley Davidson Ultimate Chopper', 'Motorcycles', '1:10', 'Min Lin Diecast', 'This replica features working kickstand, front suspension, gear-shift lever, footbrake lever, drive chain, wheels and steering. All parts are particularly delicate due to their precise scale and require special care and attention.', 7933, Decimal('48.81'), Decimal('95.70')),
 ('S10_1949', '1952 Alpine Renault 1300', 'Classic Cars', '1:10', 'Classic Metal Creations', 'Turnable front wheels; steering function; detailed interior; detailed engine; opening hood; opening trunk; opening doors; and detailed chassis.', 7305, Decimal('98.58'), Decimal('214.30')),
 ('S10_2016', '1996 Moto Guzzi 1100i', 'Motorcycles', '1:10', 'Highway 66 Mini Classics', 'Official Moto Guzzi logos and insignias, saddle bags located on side of motorcycle, detailed engine, working steering, working suspension, two leathe ... (93 characters truncated) ... paint with chrome accents, superior die-cast detail , rotating wheels , working kick stand, diecast metal with plastic parts and baked enamel finish.', 6625, Decimal('68.99'), Decimal('118.94')),
 ('S10_4698', '2003 Harley-Davidson Eagle Drag Bike', 'Motorcycles', '1:10', 'Red Start Diecast', 'Model features, official Harley Davidson logos and insignias, detachable rear wheelie bar, heavy diecast metal with resin parts, authentic multi-colo ... (199 characters truncated) ... t replica, baked enamel finish, 1:10 scale model, removable fender, seat and tank cover piece for displaying the superior detail of the v-twin engine', 5582, Decimal('91.02'), Decimal('193.66')),
 ('S10_4757', '1972 Alfa Romeo GTA', 'Classic Cars', '1:10', 'Motor City Art Classics', 'Features include: Turnable front wheels; steering function; detailed interior; detailed engine; opening hood; opening trunk; opening doors; and detailed chassis.', 3252, Decimal('85.68'), Decimal('136.00')),
 ('S10_4962', '1962 LanciaA Delta 16V', 'Classic Cars', '1:10', 'Second Gear Diecast', 'Features include: Turnable front wheels; steering function; detailed interior; detailed engine; opening hood; opening trunk; opening doors; and detailed chassis.', 6791, Decimal('103.42'), Decimal('147.74')),
 ('S12_1099', '1968 Ford Mustang', 'Classic Cars', '1:12', 'Autoart Studio Design', 'Hood, doors and trunk all open to reveal highly detailed interior features. Steering wheel actually turns the front wheels. Color dark green.', 68, Decimal('95.34'), Decimal('194.57')),
 ('S12_1108', '2001 Ferrari Enzo', 'Classic Cars', '1:12', 'Second Gear Diecast', 'Turnable front wheels; steering function; detailed interior; detailed engine; opening hood; opening trunk; opening doors; and detailed chassis.', 3619, Decimal('95.59'), Decimal('207.80')),
 ('S12_1666', '1958 Setra Bus', 'Trucks and Buses', '1:12', 'Welly Diecast Productions', 'Model features 30 windows, skylights & glare resistant glass, working steering system, original logos', 1579, Decimal('77.90'), Decimal('136.67')),
 ('S12_2823', '2002 Suzuki XREO', 'Motorcycles', '1:12', 'Unimax Art Galleries', 'Official logos and insignias, saddle bags located on side of motorcycle, detailed engine, working steering, working suspension, two leather seats, lu ... (82 characters truncated) ... paint with chrome accents, superior die-cast detail , rotating wheels , working kick stand, diecast metal with plastic parts and baked enamel finish.', 9997, Decimal('66.27'), Decimal('150.62'))]

- Just for verification, let's find the set of rows we plan to update.

In [4]:
%sql SELECT * FROM classicmodels.products where productLine='Motorcycles' and productScale='1:10';

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


[('S10_1678', '1969 Harley Davidson Ultimate Chopper', 'Motorcycles', '1:10', 'Min Lin Diecast', 'This replica features working kickstand, front suspension, gear-shift lever, footbrake lever, drive chain, wheels and steering. All parts are particularly delicate due to their precise scale and require special care and attention.', 7933, Decimal('48.81'), Decimal('95.70')),
 ('S10_2016', '1996 Moto Guzzi 1100i', 'Motorcycles', '1:10', 'Highway 66 Mini Classics', 'Official Moto Guzzi logos and insignias, saddle bags located on side of motorcycle, detailed engine, working steering, working suspension, two leathe ... (93 characters truncated) ... paint with chrome accents, superior die-cast detail , rotating wheels , working kick stand, diecast metal with plastic parts and baked enamel finish.', 6625, Decimal('68.99'), Decimal('118.94')),
 ('S10_4698', '2003 Harley-Davidson Eagle Drag Bike', 'Motorcycles', '1:10', 'Red Start Diecast', 'Model features, official Harley Davidson logos and insignias, detachable rear wheelie bar, heavy diecast metal with resin parts, authentic multi-colo ... (199 characters truncated) ... t replica, baked enamel finish, 1:10 scale model, removable fender, seat and tank cover piece for displaying the superior detail of the v-twin engine', 5582, Decimal('91.02'), Decimal('193.66'))]

- Now perform the update:

In [5]:
%%sql
update classicmodels.products set MSRP=(1.05*MSRP) where productLine='Motorcycles' and productScale='1:10';

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


C:\Users\dferguso\Anaconda\lib\site-packages\pymysql\cursors.py:170: Warning: (1265, "Data truncated for column 'MSRP' at row 1")
  result = self._query(query)
C:\Users\dferguso\Anaconda\lib\site-packages\pymysql\cursors.py:170: Warning: (1265, "Data truncated for column 'MSRP' at row 2")
  result = self._query(query)
C:\Users\dferguso\Anaconda\lib\site-packages\pymysql\cursors.py:170: Warning: (1265, "Data truncated for column 'MSRP' at row 3")
  result = self._query(query)


[]

In [6]:
%sql SELECT * FROM classicmodels.products where productLine='Motorcycles' and productScale='1:10';

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


[('S10_1678', '1969 Harley Davidson Ultimate Chopper', 'Motorcycles', '1:10', 'Min Lin Diecast', 'This replica features working kickstand, front suspension, gear-shift lever, footbrake lever, drive chain, wheels and steering. All parts are particularly delicate due to their precise scale and require special care and attention.', 7933, Decimal('48.81'), Decimal('100.49')),
 ('S10_2016', '1996 Moto Guzzi 1100i', 'Motorcycles', '1:10', 'Highway 66 Mini Classics', 'Official Moto Guzzi logos and insignias, saddle bags located on side of motorcycle, detailed engine, working steering, working suspension, two leathe ... (93 characters truncated) ... paint with chrome accents, superior die-cast detail , rotating wheels , working kick stand, diecast metal with plastic parts and baked enamel finish.', 6625, Decimal('68.99'), Decimal('124.89')),
 ('S10_4698', '2003 Harley-Davidson Eagle Drag Bike', 'Motorcycles', '1:10', 'Red Start Diecast', 'Model features, official Harley Davidson logos and insignias, detachable rear wheelie bar, heavy diecast metal with resin parts, authentic multi-colo ... (199 characters truncated) ... t replica, baked enamel finish, 1:10 scale model, removable fender, seat and tank cover piece for displaying the superior detail of the v-twin engine', 5582, Decimal('91.02'), Decimal('203.34'))]

- The truncated warning is due to the fact that:
    - The MSRP columns is of type DECIMAL(10,2).
    - Multiplying by 1.05 produces real numbers with more than two digits after the decimal. 
    - The value is truncated to DECIMAL(10,2)
    

- We can fix several ways. The easiest way is to round the result to two decimal places.


In [8]:
%%sql
update classicmodels.products set MSRP=round((1.05*MSRP),2) where productLine='Motorcycles' and productScale='1:10';

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


[]

In [10]:
%sql SELECT * FROM classicmodels.products where productLine='Motorcycles' and productScale='1:10';

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
3 rows affected.


[('S10_1678', '1969 Harley Davidson Ultimate Chopper', 'Motorcycles', '1:10', 'Min Lin Diecast', 'This replica features working kickstand, front suspension, gear-shift lever, footbrake lever, drive chain, wheels and steering. All parts are particularly delicate due to their precise scale and require special care and attention.', 7933, Decimal('48.81'), Decimal('105.51')),
 ('S10_2016', '1996 Moto Guzzi 1100i', 'Motorcycles', '1:10', 'Highway 66 Mini Classics', 'Official Moto Guzzi logos and insignias, saddle bags located on side of motorcycle, detailed engine, working steering, working suspension, two leathe ... (93 characters truncated) ... paint with chrome accents, superior die-cast detail , rotating wheels , working kick stand, diecast metal with plastic parts and baked enamel finish.', 6625, Decimal('68.99'), Decimal('131.13')),
 ('S10_4698', '2003 Harley-Davidson Eagle Drag Bike', 'Motorcycles', '1:10', 'Red Start Diecast', 'Model features, official Harley Davidson logos and insignias, detachable rear wheelie bar, heavy diecast metal with resin parts, authentic multi-colo ... (199 characters truncated) ... t replica, baked enamel finish, 1:10 scale model, removable fender, seat and tank cover piece for displaying the superior detail of the v-twin engine', 5582, Decimal('91.02'), Decimal('213.51'))]

### DELETE

#### Overview

- Very similar to ```SELECT``` and ```UPDATE.```

```
DELETE FROM table_name WHERE condition;
```

- The more complex, full MySQL syntax is:

```
DELETE [LOW_PRIORITY] [QUICK] [IGNORE] FROM tbl_name
    [PARTITION (partition_name [, partition_name] ...)]
    [WHERE where_condition]
    [ORDER BY ...]
    [LIMIT row_count]
```

#### Example

- I do not want to delete, or try to delete, too much of the various databases I have installed.


- I am going to delete a couple of the rows that I previously inserted.


In [17]:
%sql select * from classicmodels.offices

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
9 rows affected.


[('1', 'San Francisco', '+1 650 219 4782', '100 Market Street', 'Suite 300', 'CA', 'USA', '94080', 'NA'),
 ('2', 'Boston', '+1 215 837 0825', '1550 Court Place', 'Suite 102', 'MA', 'USA', '02107', 'NA'),
 ('3', 'NYC', '+1 212 555 3000', '523 East 53rd Street', 'apt. 5A', 'NY', 'USA', '10022', 'NA'),
 ('4', 'Paris', '+33 14 723 4404', "43 Rue Jouffroy D'abbans", None, None, 'France', '75017', 'EMEA'),
 ('5', 'Tokyo', '+81 33 224 5000', '4-1 Kioicho', None, 'Chiyoda-Ku', 'Japan', '102-8578', 'Japan'),
 ('6', 'Sydney', '+61 2 9264 2451', '5-11 Wentworth Avenue', 'Floor #2', None, 'Australia', 'NSW 2010', 'APAC'),
 ('7', 'London', '+44 20 7877 2041', '25 Old Broad Street', 'Level 7', None, 'UK', 'EC2N 1HN', 'EMEA'),
 ('8', 'New York', '+1 212 555-1212', 'W 116th Street', 'and Broadway', 'NY', 'USA', '10027', 'NA'),
 ('9', 'Boston', '+1 617 555 1212', '1 Government Plaza', None, None, 'USA', '02101', 'NA')]

- Just to show a slightly more complex example of ```DELETE```

In [19]:
%%sql
delete from classicmodels.offices where city='Boston' and addressLine1='1 Government Plaza'

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[]

In [21]:
%sql select * from classicmodels.offices where officeCode=9

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
0 rows affected.


[]

## HW1 Checkpoint

### Core Concept

- ```RDBDataTable``` and ```CSVDataTable``` have four core methods:
    - ```find_by_template(self, template, field_list=None, limit=None, offset=None, order_by=None)```
    - ```delete_by_template(self, template)```
    - ```update_by_template(self, template, new_values)```
    - ```insert(self, new_record)```
    

- ```by_template``` is core to three of the methods: ```find_by_template, udate_by_template, delete_by_template.```


- How do you handle templates?


### CSV Files

```1.``` The first step is to load the CSV file into a data structure, e.g. _list_


In [22]:
import csv

_default_directory = \
        "C:\\Users\\dferguso\\ansys_one_drive\\Columbia\Projects\\W4111\\HW1Impl\\Data" + \
            "\\baseballdatabank-2019.2\\core\\"

def load_csv_file(fn):
    
    # Open the file in read model
    with open(fn, "r") as in_file:
        
        # Wrap with class that helps reading CSV files.
        csv_rdr = csv.DictReader(in_file)
        
        # Place to hold the result
        result = []
        
        # Build the result
        for r in csv_rdr:
            result.append(r)
            
    return result
        
    

In [24]:
import os
fn = os.path.join(_default_directory, "People.csv")
the_people = load_csv_file(fn)

In [25]:
print("Load got", len(the_people), "rows.")

Load got 19617 rows.


In [28]:
import json
print("A typical row looks like: \n", json.dumps(the_people[0], indent=2))

A typical row looks like: 
 {
  "playerID": "aardsda01",
  "birthYear": "1981",
  "birthMonth": "12",
  "birthDay": "27",
  "birthCountry": "USA",
  "birthState": "CO",
  "birthCity": "Denver",
  "deathYear": "",
  "deathMonth": "",
  "deathDay": "",
  "deathCountry": "",
  "deathState": "",
  "deathCity": "",
  "nameFirst": "David",
  "nameLast": "Aardsma",
  "nameGiven": "David Allan",
  "weight": "215",
  "height": "75",
  "bats": "R",
  "throws": "R",
  "debut": "2004-04-06",
  "finalGame": "2015-08-23",
  "retroID": "aardd001",
  "bbrefID": "aardsda01"
}


```2.``` How do you check if a specific row matches a template?

In [34]:
template = {"nameLast": "Aardsma", "throws": "R"}


def matches_template(row, template):
    
    result = True
    
    # Look at every attribute in the template and see if there is a match.
    for k,v in template.items():
        
        # Get the corresponding value in the row, or None if not there.
        row_v = row.get(k, None)
       
        if v != row_v:
            result = False
            break
            
    return result

In [35]:
match = matches_template(the_people[0], template)
match

True

In [36]:
template2 = {"nameLast": "Aardsma", "throws": "L"}
match = matches_template(the_people[0], template2)
match

False

```3.``` Apply the matcher to all the rows. This is a simple loop.


```4.``` You also need to get the subset of fields requested.

In [37]:
def get_fields(row, field_list):
    
    result = {}
    for f in field_list:
        
        # A good program would catch an exception if field is not in the row.
        # Do not worry about it. This is not a robust programming course.
        result[f] = row[f]
        
    return result

In [38]:
field_list = ['nameFirst', 'nameLast', 'birthState']
requested_fields = get_fields(the_people[0], field_list)
print("Requested fields = ", json.dumps(requested_fields, indent=2))

Requested fields =  {
  "nameFirst": "David",
  "nameLast": "Aardsma",
  "birthState": "CO"
}


### How do You do this for SQL

```1.``` You have to make a ```WHERE``` clause.

In [46]:
def template_to_where_clause(template):
    
    if template is None or template == {}:
        w_clause = None
        args = None
    else:
        terms = []
        args = []
        for k,v in template.items():
            terms.append(k + "=%s")
            args.append(v)
            
        w_clause = "where " + (" and ".join(terms))
        
    return w_clause, args

In [47]:
template = {"nameLast": "Aardsma", "nameFirst": "David"}
w_clause, args = template_to_where_clause(template)

In [48]:
w_clause

'where nameLast=%s and nameFirst=%s'

In [49]:
args

['Aardsma', 'David']

```2.``` You also have to pick some fields.

In [51]:
def get_select_fields(fields):
    
    if fields is None or fields == []:
        field_list = " * "
    else:
        field_list = ",".join(fields)
        
    return field_list

In [52]:
fields = ['nameLast', 'nameFirst', 'birthCity', 'throws']
field_list = get_select_fields(fields)
field_list

'nameLast,nameFirst,birthCity,throws'

```3.``` Now, we can put the pieces together.

In [56]:
fields = ['nameLast', 'nameFirst', 'birthCity', 'throws']
template = {"nameLast": "Aardsma", "nameFirst": "David"}
w_clause, arhs = template_to_where_clause(template)
field_list = get_select_fields(fields)
table_name = "lahman2019raw.people"

In [63]:
sql = "select " + field_list + " from " + table_name + " " + w_clause
sql

'select nameLast,nameFirst,birthCity,throws from lahman2019raw.people where nameLast=%s and nameFirst=%s'

```4.``` Now run the query.

In [64]:
cur = default_cnx.cursor()
res = cur.execute(sql, args)
result = cur.fetchall()
print("Result = ", json.dumps(result, indent=2))

Result =  [
  {
    "nameLast": "Aardsma",
    "nameFirst": "David",
    "birthCity": "Denver",
    "throws": "R"
  }
]


### Summary

1. This HW may seem overwhelming. It is actually very simple.<br><br>
2. The homework only requires programming knowledge that you learn in 1004, 1007 or 1007.<br><br>
3. Solving the HW requires:
    - Writing several "helper" functions each of which does a simple task.
    - Putting the smaller pieces together into the larger methods in the class.<br><br>
4. The complexity stems for the fact that we do not sufficiently teach you how to write larger programs through decomposition and assembly. All of your past programming assignments have not taught you do develop and test this way.

## Primary Key

### Conept

- Briefly discussed the concept in lecture 1, and will cover in more detail now.


- There are several types of keys:
    - Super key: A set of attributes within a table whose values can be used to uniquely identify a tuple. 
    - Candidate key: A minimal super key. That is, no subset of the columns is a super key.
    - Primary key: Candidate key arbitrarily chosen to be the "primary" key.
    - Alternate key: Alternate key is the key that can work as a primary key. Basically it is a candidate key that currently is not primary key.
    - Composite/Compound key: Composite Key is a combination of more than one fields/columns of a table. It can be a Candidate key, Primary key.
    - Unique Key: Unique key is a set of one or more fields/columns of a table that uniquely identify a record in database table but can be NULL.
    - Foreign key: Foreign key is a a set of columns in database table that is primary key in another table. It can accept multiple null, duplicate values.
    - Natural key: A natural key (also known as business key) is a type of unique key, found in relational model database design, that is formed of attributes that already exist in the real world. It is used in business-related columns. In other words, a natural key is a candidate key that has a logical relationship to the attributes within that row. A natural key is sometimes called domain key. (https://en.wikipedia.org/wiki/Candidate_key)
    - Surrogate key: A surrogate key (or synthetic key, entity identifier, system-generated key, database sequence number, factless key, technical key, or arbitrary unique identifier) in a database is a unique identifier for either an entity in the modeled world or an object in the database. The surrogate key is not derived from application data, unlike a natural (or business) key which is derived from application data. (https://en.wikipedia.org/wiki/Surrogate_key)


- "In the relational model of databases, a primary key is a specific choice of a minimal set of attributes (columns) that uniquely specify a tuple (row) in a relation (table). Informally, a primary key is 'which attributes identify a record', and in simple cases are simply a single attribute: a unique id. More formally, a primary key is a choice of candidate key (a minimal superkey); any other candidate key is an alternate key.<br><br>
A primary key may consist of real-world observables, in which case it is called a natural key, while an attribute created to function as a key is called a surrogate key. For example, for a database of people (of a given nationality), time and location of birth is a natural key, while national identification numbers are surrogate keys." (https://en.wikipedia.org/wiki/Primary_key)


- __Note:__
    - Time and location of birth is a terrible natural key.
    - A better example is ```Appearances``` in Lahman database, the columns: ```playerID, teamID, yearID``` uniquely identify a tuple.
    - The RDB engine will enforce any defined primary key. Your code just needs to handle any exceptions.
    - You will need to implement primary key definition on ```CSVDataTable``` instance creation.


### Relational Model $-$ Keys

- "_Data integrity_ is the maintenance of, and the assurance of the accuracy and consistency of, data over its entire life-cycle, and is a critical aspect to the design, implementation and usage of any system which stores, processes, or retrieves data." (https://en.wikipedia.org/wiki/Data_integrity)


- _Relational (integrity) constraints_ are central to the value of relational databases.
    - The database designer _declares (defines)_ integrity constraints.
    - The database management system _rejects_ any _create, update_ or _delete_ operation that would result in a constraint violation.


- _Key(s)_ is a core  _(integrity) constraint_ enforcing _data integrity._ 
    - A _super key_ is a combination of columns with the property that no two rows have the same values for the fields of a super key.
    - A _candidate key_ is a minimal _super key_, that is removing a column from the key definition means that the key no longer uniquely identifies a row.,
    - The _primary key_ is a candidate key subjectively chosen as the "best key" for uniquely identifying the tuples.
    
    
- You will some times here the term _functionally determines._
    - Assume we have a relation $R(a,b,c,d,e)$.
    - If $(a,b)$ is a key $k$, then
    - The we can say that $k(a,b)$ _functionally determines_ $(c,d,e).$
    - Given values $(x,y)$ for $(a,b)$ we can functionally return $(c,d,e).$ The function is
    
    $\pi$<sub>$c,d,e$</sub>$(\sigma$<sub>$(a=x)\land(b=y)$</sub>$(R))$
    

- One common notation for identifying a key in the relational notation is an ```overline.```<br>      
\begin{equation}
R = (\overline{a,b},c,d)
\end{equation}

### Keys and Indexes $-$ SQL

#### Overview

- Relational database engines support additional key semantics and also support _indexes._


- MySQL supports the following keys/index:
    - PRIMARY KEY: Key is unique and NOT NULL. $\Rightarrow$ Indexed
    - UNIQUE: Key is unique or NULL. Multiple rows may have NULL for a value if column definition allows NULL. $\Rightarrow$ Indexed
    - INDEX: Non-unique but indexed for performance.
    - FULLTEXT (INDEX) -- Will cover later.
    
    
- In database engines, keys represent/implement
    - Integrity constraints. We will cover the concept in detail in future lectures.
    - Performance optimization.
    
    
- For now, just think:
    - Every table should have a primary key.
    - Put indexes on other columns or sets of columns that will frequently be in WHERE clauses.
    
    
- For a table with 1,000,000 rows:
    - SELECT without a index examines O(1,000,000) rows.
    - SELECT with an index examines O(log(1,000,000)) = 20 rows.
    

- RDB query engine compute a "cost" in terms of CPU and disk input/output for relational operations. This is critical to _query optimization,_ which we will cover later. For comparison of the benefits of indexes, the cost of ```select * from people where playerid='willite01'```
    - Without index = 4,071.
    - With index = 1
    
    
- We will cover query optimization and cost in Module II: Database Management System Implementation/Architecture.

#### An Example

- A course directory entry for a database course/section.

<hr style="height: 1px;">

| <img src="../../images/course_directory_entry.jpeg"> |
| :---: |
| __Student SSOL for W4111__ |

<hr style="height: 1px;">


- I pulled down some course data for me.


- How do I figure out the keys?


- You can use ```SELECT COUNT(*) GROUP BY A,B,C``` for various sets of columns.


- We will cover _Aggregation_ in the next lecture, but you can use now.


- Let's look at the data.

In [68]:
%sql select * from w4111f19.course_catalog

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
15 rows affected.


[('11912', 'TOPICS IN COMPUTER SCIENCE', 'COMSE6998', '010', '2016', '3'),
 ('34758', 'TOPICS IN COMPUTER SCIENCE', 'COMSE6998', '006', '2017', '1'),
 ('47421', 'TOPICS IN COMPUTER SCIENCE', 'COMSE6998', '014', '2017', '3'),
 ('47468', 'INTRODUCTION TO DATABASES', 'COMSW4111', '003', '2017', '3'),
 ('53309', 'INTRODUCTION TO DATABASES', 'COMSW4111', '002', '2018', '1'),
 ('53496', 'INTRO TO COMP FOR ENG/APP SCI', 'ENGIE1006', '001', '2018', '1'),
 ('59651', 'FIELDWORK', 'COMSW3999', '063', '2018', '3'),
 ('59653', 'INTRODUCTION TO DATABASES', 'COMSW4111', '001', '2018', '3'),
 ('61720', 'THE ART OF ENGINEERING', 'ENGIE1102', '001', '2018', '3'),
 ('69174', 'TOPICS IN SOFTWARE ENGINEERING: Cloud and Microservice Applications', 'COMSE6156', '001', '2018', '3'),
 ('79571', 'INTRODUCTION TO DATABASES', 'COMSW4111', '003', '2019', '1'),
 ('82004', 'FIELDWORK', 'COMSW3999', '063', '2019', '2'),
 ('85777', 'INTRODUCTION TO DATABASES', 'COMSW4111', '002', '2019', '3'),
 ('87722', 'TOPICS IN SOFTWARE ENGINEERING', 'COMSE6156', '001', '2019', '3'),
 ('91884', 'INTRODUCTION TO DATABASES', 'COMSW4111', 'H02', '2019', '3')]

In [69]:
%sql select count(*) as no_of_rows from w4111f19.course_catalog

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
1 rows affected.


[(15,)]

- Is ```course_no``` a key?

In [74]:
%sql select course_no, count(*) as no_of_occurences  from w4111f19.course_catalog group by course_no;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
6 rows affected.


[('COMSE6998', 3),
 ('COMSW4111', 6),
 ('ENGIE1006', 1),
 ('COMSW3999', 2),
 ('ENGIE1102', 1),
 ('COMSE6156', 2)]

- Clearly not.

- How about ```course_no``` and ```year?```

In [76]:
%sql select course_no, year, count(*) as no_of_occurences  from w4111f19.course_catalog group by course_no, year;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
11 rows affected.


[('COMSE6998', '2016', 1),
 ('COMSE6998', '2017', 2),
 ('COMSW4111', '2017', 1),
 ('COMSW4111', '2018', 2),
 ('ENGIE1006', '2018', 1),
 ('COMSW3999', '2018', 1),
 ('ENGIE1102', '2018', 1),
 ('COMSE6156', '2018', 1),
 ('COMSW4111', '2019', 3),
 ('COMSW3999', '2019', 1),
 ('COMSE6156', '2019', 1)]

- How about ```course_no, year, section?```

In [77]:
%sql select course_no, year, section, count(*) as no_of_occurences  from w4111f19.course_catalog \
    group by course_no, year, section;

 * mysql+pymysql://dbuser:***@localhost/lahman2019raw
15 rows affected.


[('COMSE6998', '2016', '010', 1),
 ('COMSE6998', '2017', '006', 1),
 ('COMSE6998', '2017', '014', 1),
 ('COMSW4111', '2017', '003', 1),
 ('COMSW4111', '2018', '002', 1),
 ('ENGIE1006', '2018', '001', 1),
 ('COMSW3999', '2018', '063', 1),
 ('COMSW4111', '2018', '001', 1),
 ('ENGIE1102', '2018', '001', 1),
 ('COMSE6156', '2018', '001', 1),
 ('COMSW4111', '2019', '003', 1),
 ('COMSW3999', '2019', '063', 1),
 ('COMSW4111', '2019', '002', 1),
 ('COMSE6156', '2019', '001', 1),
 ('COMSW4111', '2019', 'H02', 1)]

- That looks promising, but ... 
    - This is due to the fact that these are the courses that I teach.
    - There are three or four database sections each semester.
    - It is possible that I could teach ```section 003``` in different semesters. This has not happened, __but it could.__
    
    
- A key is a statement about the __all possible values for the relation/table,__ not just the rows currently in the table.


- In this simple case, the actual answer is:
    - Course number.
    - Section number.
    - Semester.
    - Year.

# Data Modeling and Some Examples

## Setup Scenario

- Let's assume we we are going to build a course/section database for Columbia courses.


- We would conduct interviews and look at forms/documentation to determine what the data model underlying the information is.


- There already are databases, but let's pretend we are trying to build the database from existing information (docs, forms).


## Course and Section Information Table

- Let's look at some of the information that we would find in interviews, looking at forms, ...

| <img src="../../images/course_info_1.jpg" width="900px"> |
| :---: |
| __Course/Section Information: Course Works__ |

| <img src="../../images/course_info_2.jpg" width="900px"> |
| :---: |
| __Course/Section Information: Section Information__ |

| <img src="../../images/course_info_3.jpg" width="900px"> |
| :---: |
| __Course/Section Information: Scheduled Classes__ |


| <img src="../../images/course_info_4.jpg" width="900px"> |
| :---: |
| __Course/Section Information: Standard Times__ |